data/tutorials/wf_03_metaprogramming.md

```diff
@@ -0,0 +1,454 @@
+more the build system's role to drive the preprocessing. So, applying
+preprocessors will boil down to telling Dune about it. Only for educational
+purposes, in the next section we show how to preprocess a file using OCaml's
+compiler, the more relevant part being [Preprocessing with
+Dune](#preprocessing-with-dune).
+
+### Preprocessing with `ocamlc` and `ocamlopt`
+
+OCaml's compiler `ocamlc` and `ocamlopt` offer the `-pp` option to preprocess a
+file in the compilation phase (but remember that you are encourage to use Dune
+to drive the preprocessing). Consider the following simple preprocessor which
+replaces the string `"World"` by the string `"Universe"`, here in the form of a
+shell script:
+
+```shell
+$ cat preprocessor.sh
+#!/bin/sh
+
+sed 's/World/Universe/g' $1
+```
+
+Compiling a classic "Hello, World!" program with this option
+would alter the compilation:
+
+```shell
+$ cat hello.ml
+print_endline "Hello, World!";;
+
+$ ocamlopt -pp preprocessor.sh hello.ml
+$ ./a.out
+Hello, Universe!
+```
+
+### Preprocessing with Dune
+
+Dune's build system has a specific stanza to apply preprocessing to files.
+The full documentation can be found
+[here](https://dune.readthedocs.io/en/stable/concepts.html#preprocessing-specification),
+and should serve as a reference. In this section, we only give a few examples.
+
+The stanza to apply preprocessing on the source file is `(preprocess (action
+(<action>)))`. The `<action>` part can be any user-defined action, such as the
+call to an external program, as specified by `(system <program>)`.
+
+Putting it all together, the following `dune` file would rewrite the corresponding
+module files using our previously written `preprocessor.sh`:
+
+```dune
+(executable
+ (name main)
+ (preprocess
+  (action
+   (system "./preprocessor.sh %{input-file}"))))
+```
+
+### The Limits of Manipulating Text Files
+
+The complexities of a programming language syntax makes it very hard to
+manipulate text in a way that is tied to the syntax. Suppose for instance that,
+similarly to the previous example, you want to rewrite all occurences of "World"
+by "Universe," but _inside the OCaml strings of the program_ only. It is quite
+involved and requires a good knowledge of the OCaml syntax to do so, as there
+are several delimiters for strings (such as `{| ...|}`) and line breaks, or
+comments could get in the way...
+
+Consider another example. Suppose you have defined a type, and you want to
+generate a serializer at compile time from this specific type to an encoding of
+it in a `json` format, such as `Yojson` (see
+[here](#why-are-ppxes-especially-useful-in-ocaml) for the reasons it has to be
+generated at compile-time). This serialization code could be written by a
+preprocessor, which would look for the type in the file and serialize it
+differently depending on the type structure; that is, whether it is a variants
+type, a record type, the structure of its subtypes...
```

Comment:
(I admit that the last point is only slightly related to the AST having a richer structure)

Reply:
The thing is I wanted to avoid explicitly speaking about the AST, as it has not yet been introduced. I tried to basically say the thing with the first sentence ("The complexities of a programming language syntax makes it very hard to manipulate text in a way that is tied to the syntax.") but this might not be enough!

About your comments:
> - there's no "AST type" safety: you can add source code corresponding to an AST node of type B at a place where the parser expects an AST node of type A

I think this is very convincing: you _cannot_ generate code with syntax-error with PPX, while you can with text preprocessors. However, that's not easy to explain without speaking about AST!
> - you can introduce (other) syntax errors via a source code preprocessor, which are always hard to debug

I am not sure to see the difference with the first point! (And with PPXs you can still generate errors discovered later by the compiler such as type errors, which can also be difficult to debug! For instance for some reason I had the corresponding warning turned on and a ppx opened unused module, it took me some time to understand what was happening...)

> it's always global. therefore
    it doesn't compose well with other source code preprocessors

I'll add that!

>    it's inefficient

I don't think it is inherent to a text preprocessing system to be inefficient... 

I will push another take on this in a couple of minutes, let me know what you think!

Reply:
> The thing is I wanted to avoid explicitly speaking about the AST, as it has not yet been introduced.

Ok, that seems like a fair choice of style! Just to mention: another style that might also fit is the following: a fundamental concept (such as AST in this doc) is mentioned already at the beginning. It is then given more shape step by step throughout the document before being explained in detail. That way, the concept becomes clearer first on a high level (it is an abstraction of the language; it has far more structure than the source code etc.) and later it's explained on a low level (the way you've explained it: it consists of a list of structure items etc). But you've been writing this, so of course, the choice is yours!

Reply:
> I am not sure to see the difference with the first point!

The difference is just that in the first point I was talking about a concrete kind of syntax error and in the second point, I mentioned that with source code preprocessing you could generate _any_ kind of syntax errors, e.g. something as simple as a typo.

Reply:
> I don't think it is inherent to a text preprocessing system to be inefficient...

Do you think text preprocessing allows local preprocessing? What I implicitly meant was that with text preprocessing, you couldn't join all local (i.e. ideally pretty much all) preprocessors into one pass (the way the ppxlib driver does with context-free rules). Btw, have I also already mentioned bad composition semantics as a down-side of text preprocessors?

Reply:
> Do you think text preprocessing allows local preprocessing?

Yes! Tokens such as `blibli blibli MACRO_TO_BE_REPLACED(payload) blablab` are replaced in a context-free way. On could have a "ppxlib for text preprocessing" where you register rules on how to replace macros and it generates a single executable expanding the macro.

> Btw, have I also already mentioned bad composition semantics as a down-side of text preprocessors?

I don't think that bad composition semantics is a downside of "text preprocessing", but more a downside of "no macro language".

Indeed, with ppx (but without ppxlib) it is exactly the same thing: order matter. And similarly, with a "ppxlib for text", there would be no composition issue. What ppxlib is doing I think is "replacing" a macro language, puting itself inbetween transformations (written using a specific language: extenders and derivers).

So, we cannot really add this in the limits of preprocessing text files. However, I emphasized it in the advantages of using ppxlib rather than implementing it your own way.